src/runtime/crt/common/crt_runtime_api.c

```diff
@@ -38,7 +38,10 @@
 
 static char g_last_error[1024];
 
-void TVMAPISetLastError(const char* msg) { strncpy(g_last_error, msg, sizeof(g_last_error)); }
+void TVMAPISetLastError(const char* msg) {
+  strncpy(g_last_error, msg, sizeof(g_last_error) - 1);
+  g_last_error[sizeof(g_last_error) - 1] = 0;
```

Comment:
I think no null-character is implicitly appended at the end of destination only if source is longer than num, otherwise, destination is padded with zeros until a total of num characters have been written to it. See [strncpy - C++ reference](http://www.cplusplus.com/reference/cstring/strncpy/).

The error reported in the compiler meant "bound 1024 equals destination size". As a correction for this, I think it'll be better to copy msg directly to g_last_error:
```c
strcpy(g_last_error, msg);
```

Reply:
Does "truncate msg to the size of g_last_error" mean mutating "msg"? That doesn't sound always doable. My understanding:

1. Before the proposed change, when the error message is longer than 1024, g_last_error won't be null terminated. I believe this is what the warning is about.
2. With the change, g_last_error will always be null-terminated.
3. If we use `strcpy`, we need to strlen first. If the length is long, we need to do something special. I don't think it's trivial. It could be faster, but I wonder how much it matters, since errors are probably rare.
4. snprintf may be an alternative. But probably not too different in the context.

Given these, would you reconsider the change as is?

Reply:
Good point. On a second thought, the proposed change is good enough to put msg to g_last_error.